src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11664,14 +11664,28 @@ SyntaxKind.PlusToken or
                         SyntaxKind.MinusToken or
                         SyntaxKind.AmpersandToken or
                         SyntaxKind.AsteriskToken or
-                        SyntaxKind.DotDotToken => true,
-                        var tk => CanFollowCast(tk)
+                        SyntaxKind.DotDotToken or
+                        SyntaxKind.OpenBracketToken // for `(int[])[1, 2, 3]` or `(List<int>?)[1, 2, 3]`
+                            => true,
+
+                        var tk
+                            => CanFollowCast(tk)
                     };
 
                 case ScanTypeFlags.GenericTypeOrMethod:
+                case ScanTypeFlags.TupleType:
+                    // If we have `(X<Y>)[...` then we know this must be a cast of a collection literal, not an index
+                    // into some expr. As most collections are generic this means a good experience most of the time in
+                    // this ambiguous scenario.
```

Comment:
This sentence seems to indicate the generic case is ambiguous, even though the previous sentence implies it is not. Would it make sense to reword this second sentence? Perhaps: "As most collections are generic, the common case is not ambiguous."